LogEventInfo.MessageTemplate - Subset of LogEventInfo.Properties

snakefoot opened the pull request:
Ordered by position (instead of hashcode).

It is possible to override these properties using the standard LogEventInfo-Properties-IDictionary-interface.

Allows capture of Microsoft FormattedLogValues without loosing the order given in the MessageTemplate. https://github.com/NLog/NLog.Extensions.Logging/pull/125

snakefoot commented:
@304NotModified Have fixed merge conflicts, so we can continue the journey.

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+            }
+        }
+
+        private int MessagePropertiesIndexOfKey(object key)
```
proposal name: GetMessagePropertiesIndex

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+            {
+                if (_eventProperties != null && (_messageProperties == null || _messageProperties.Count == 0))
+                    return _eventProperties.Keys;
+                if (_eventProperties == null && (_messageProperties == null || _messageProperties.Count == 0))
```
I prefer an `else` and nested `if`s here. 

Maybe also a helper for `_messageProperties == null || _messageProperties.Count == 0` (9 times used?)

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+        /// <inheritDoc/>
+        public bool Remove(object key)
+        {
+            if (_eventProperties != null && _eventProperties.Remove(key))
```
prefer `return (_eventProperties != null && _eventProperties.Remove(key))`

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+                {
+                    if (_eventEnumeratorCreated)
+                    {
+                        if (_eventEnumerator.MoveNext())
```
idem, `return _eventEnumerator.MoveNext()`

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+                                return true;
+                        }
+                    }
+                    if (_dictionary._eventProperties != null && _dictionary._eventProperties.ContainsValue(item))
```
idem, `return...`

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+        {
+            int keyIndex = MessagePropertiesIndexOfKey(key);
+            if (keyIndex != -1)
+                throw new ArgumentException("duplicate", "key");
```
I think it would be nice to also include the duplicate key in the exception. We can do that in the exceptiondata or the message

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+            return -1;
+        }
+
+        private static IList<KeyValuePair<string, object>> MakePropertiesUnique(IList<KeyValuePair<string, object>> value)
```
please private methods below the publics

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+        /// <inheritDoc/>
+        public void Add(object key, object value)
+        {
+            int keyIndex = MessagePropertiesIndexOfKey(key);
```
so add is now `O(n)`?

snakefoot replied:
Only when using structured-logging, and there will only be a performance hit if using more than 100 message-parameters.

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,698 @@
+        /// <inheritDoc/>
+        public void Add(KeyValuePair<object, object> item)
+        {
+            int keyIndex = MessagePropertiesIndexOfKey(item.Key);
```
call the other `add` method?

snakefoot replied:
Fixed

snakefoot commented:
> Al lot of actions are now O(n)? - which is a bit unfortunate, what the use of the dictionary then ;)

As long the number for MessageTemplateProperties is below 100 parameters, then you will not notice O(n). Normal Properties will still work as O(1).

> Correct me if I didn't understand it correctly.

I want to have a property that works just like Microsofts FormattedLogValues, Serilogs MessageTemplate.Tokens / PropertyBinder.ConstructProperties or NLog StructuredEvents Template.Holes.

And using a standard array is much more optimal memory-wise than a dynamic container like a dictionary. Since we have to merge two containers (user-defined-properties + message-defined-properties). Then I suggest that the message-defined-properties uses the most optimal sub-container.

snakefoot commented:
@304NotModified Alternative then we should just leave the Properties-dictionary alone, and just create a new merged-readonly-view of Properties and MessageTemplateProperties. Then just change all locations to use the merged-readonly-view by default instead of the Properties-dictionary.

snakefoot commented:
@304NotModified Changed the PropertiesDictionary so it always allocates and uses the dictionary. This gives extra allocation for structured-logging, but gives O(1) lookup speed for both property-types.

304NotModified commented:
Thanks! These kind of PRs are a bit hard to review, but will try to look at it soon.

snakefoot commented:
> Thanks! These kind of PRs are a bit hard to review, but will try to look at it soon.

It is a key-stone for the implementation of structured logging, so it mostly depends on how fast you want the structured logging up and running for NLog ver. 4.5.

304NotModified commented:
Maybe a stupid question (and we discussed it before long time ago i'm afraid),

Why do we need a separate collection for the messageProperties? Why isn't the current `logeventInfo.Properties` not enough? AFAIK we are missing the following features when using `logeventInfo.Properties`:

- the ordering of the messageProperties is lost
- the info of what is from the template and what not is lost.

I prefer adding all properties to the current properties collection and maybe add some more metadata if we need it for the ordering.

snakefoot commented:
> I prefer adding all properties to the current properties collection and maybe add some more metadata if we need it for the ordering.

That is what the PropertiesDictionary is doing.

304NotModified commented:
> That is what the PropertiesDictionary is doing.

true, but i'm not keen on the `logEventInfo.MessageTemplateProperties` option. 

> Why do we need a separate collection for the messageProperties? Why isn't the current logeventInfo.Properties not enough? AFAIK we are missing the following features when using (only) logeventInfo.Properties:
> 
> the ordering of the messageProperties is lost
> the info of what is from the template and what not is lost.

Is this sound and complete?

snakefoot commented:
> true, but i'm not keen on the logEventInfo.MessageTemplateProperties option.

You are welcome to suggest a different name, or a different approach for providing the messagetemplate-properties in the correct order.

The ordering is important, when logging to Json or a no-sql-database.

snakefoot commented:
@304NotModified We could also make the parsed MessageTemplate fully public (Like you did for NLog 4.5 Alpha) if that is more acceptable. Could properly make it so it is just a single array allocation of PropertyEntries-struct (Name + Value + Format)

304NotModified commented:
> @304NotModified Else we can replace MessageTemplateProperties with a public method called LogEventInfo.AddProperties(IList<KeyValuePair<string, object>> properties) ?

Yes, Something like that. I'm now creating a small (non-code) proposal



> I'm guessing 9 out of 10 doesn't use a combination of structured-logging-properties and custom-properties. And

I think also this is true

> @304NotModified We could also make the parsed MessageTemplate fully public (Like you did for NLog 4.5 Alpha) if that is more acceptable.

I like this, but I could understand this is a memory eater. Less keep this as plan B

> We can then consider to give a small life-boat, by extending the different property-layoutrenders to dive in and only output the property-subset found in PositionDictionary.MessageProperties

Could you elaborate on this? Which layout renderers should we change? I think `${event-properties}` and `${all-event-properties}`?

snakefoot commented:
> Could you elaborate on this? Which layout renderers should we change? I think ${event-properties} and ${all-event-properties}?

Yes, and also JsonLayout.IncludeAllProperties and Log4JXmlEventLayout.IncludeAllProperties

304NotModified commented:
I was thinking of something like this:

![image](https://user-images.githubusercontent.com/5808377/30664312-8898eeb6-9e4d-11e7-8e18-fe9b0a637a5a.png)

In code:

```c#
    public interface IProperties : IDictionary<object, object>
    {
        /// <summary>
        /// Add propery to the end of the collection 
        /// </summary>
        /// <param name="key"></param>
        /// <param name="value"></param>
        void AddOrderedPropery(string key, object value);

        /// <summary>
        /// Add properies to the end of the collection 
        /// </summary>
        /// <param name="properties"></param>
        void AddOrderedProperies(IEnumerable<KeyValuePair<string, object>> properties);

        /// <summary>
        /// Remove ordered property by key
        /// </summary>
        /// <param name="key"></param>
        /// <returns>found?</returns>
        bool RemoveOrderedProperty(string key);

        /// <summary>
        /// Get ordered properties in order of the template
        /// </summary>
        /// <returns></returns>
        IList<object> GetOrderedProperties();

        /// <summary>
        /// Get all properties which aren't in the template
        /// </summary>
        /// <returns></returns>
        IDictionary<object, object> GetNonOrderedProperties();

    }
```

```c#
  public class EventProperties : IProperties
    {
        /// <summary>
        /// Dict from key to index
        /// </summary>
        private Dictionary<string, int> _propertyIndexes;


        private Dictionary<object, object> _allProperties;

        public EventProperties()
        {
            _propertyIndexes = new Dictionary<string, int>(); //todo lazy init?
            _allProperties = new Dictionary<object, object>();//todo lazy init?
        }



        #region delegate _allProperties 

        /// <summary>Returns an enumerator that iterates through the collection.</summary>
        /// <returns>An enumerator that can be used to iterate through the collection.</returns>
        public IEnumerator<KeyValuePair<object, object>> GetEnumerator()
        {
            return _allProperties.GetEnumerator();
        }

        /// <summary>Returns an enumerator that iterates through a collection.</summary>
        /// <returns>An <see cref="T:System.Collections.IEnumerator" /> object that can be used to iterate through the collection.</returns>
        IEnumerator IEnumerable.GetEnumerator()
        {
            return ((IEnumerable)_allProperties).GetEnumerator();
        }

        /// <summary>Copies the elements of the <see cref="T:System.Collections.ICollection" /> to an <see cref="T:System.Array" />, starting at a particular <see cref="T:System.Array" /> index.</summary>
        /// <param name="array">The one-dimensional <see cref="T:System.Array" /> that is the destination of the elements copied from <see cref="T:System.Collections.ICollection" />. The <see cref="T:System.Array" /> must have zero-based indexing. </param>
        /// <param name="index">The zero-based index in <paramref name="array" /> at which copying begins. </param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="array" /> is null. </exception>
        /// <exception cref="T:System.ArgumentOutOfRangeException">
        /// <paramref name="index" /> is less than zero. </exception>
        /// <exception cref="T:System.ArgumentException">
        /// <paramref name="array" /> is multidimensional.-or- The number of elements in the source <see cref="T:System.Collections.ICollection" /> is greater than the available space from <paramref name="index" /> to the end of the destination <paramref name="array" />.-or-The type of the source <see cref="T:System.Collections.ICollection" /> cannot be cast automatically to the type of the destination <paramref name="array" />.</exception>
        public void CopyTo(Array array, int index)
        {
            ((ICollection)_allProperties).CopyTo(array, index);
        }

        /// <summary>Gets an object that can be used to synchronize access to the <see cref="T:System.Collections.ICollection" />.</summary>
        /// <returns>An object that can be used to synchronize access to the <see cref="T:System.Collections.ICollection" />.</returns>
        public object SyncRoot => ((ICollection)_allProperties).SyncRoot;

        /// <summary>Gets a value indicating whether access to the <see cref="T:System.Collections.ICollection" /> is synchronized (thread safe).</summary>
        /// <returns>true if access to the <see cref="T:System.Collections.ICollection" /> is synchronized (thread safe); otherwise, false.</returns>
        public bool IsSynchronized => ((ICollection)_allProperties).IsSynchronized;

        /// <summary>Determines whether the <see cref="T:System.Collections.IDictionary" /> object contains an element with the specified key.</summary>
        /// <returns>true if the <see cref="T:System.Collections.IDictionary" /> contains an element with the key; otherwise, false.</returns>
        /// <param name="key">The key to locate in the <see cref="T:System.Collections.IDictionary" /> object.</param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="key" /> is null. </exception>
        public bool Contains(object key)
        {
            return ((IDictionary)_allProperties).Contains(key);
        }

        /// <summary>Gets a value indicating whether the <see cref="T:System.Collections.IDictionary" /> object has a fixed size.</summary>
        /// <returns>true if the <see cref="T:System.Collections.IDictionary" /> object has a fixed size; otherwise, false.</returns>
        public bool IsFixedSize => ((IDictionary)_allProperties).IsFixedSize;

        /// <summary>Adds the specified key and value to the dictionary.</summary>
        /// <param name="key">The key of the element to add.</param>
        /// <param name="value">The value of the element to add. The value can be null for reference types.</param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="key" /> is null.</exception>
        /// <exception cref="T:System.ArgumentException">An element with the same key already exists in the <see cref="T:System.Collections.Generic.Dictionary`2" />.</exception>
        public void Add(object key, object value)
        {
            _allProperties.Add(key, value);
        }

        /// <summary>Adds an item to the <see cref="T:System.Collections.Generic.ICollection`1" />.</summary>
        /// <param name="item">The object to add to the <see cref="T:System.Collections.Generic.ICollection`1" />.</param>
        /// <exception cref="T:System.NotSupportedException">The <see cref="T:System.Collections.Generic.ICollection`1" /> is read-only.</exception>
        public void Add(KeyValuePair<object, object> item)
        {
            _allProperties.Add(item.Key, item.Value);
        }

        /// <summary>Removes all keys and values from the <see cref="T:System.Collections.Generic.Dictionary`2" />.</summary>
        public void Clear()
        {
            _allProperties.Clear();
        }

        /// <summary>Determines whether the <see cref="T:System.Collections.Generic.ICollection`1" /> contains a specific value.</summary>
        /// <returns>true if <paramref name="item" /> is found in the <see cref="T:System.Collections.Generic.ICollection`1" />; otherwise, false.</returns>
        /// <param name="item">The object to locate in the <see cref="T:System.Collections.Generic.ICollection`1" />.</param>
        public bool Contains(KeyValuePair<object, object> item)
        {
            throw new NotImplementedException();
        }

        /// <summary>Copies the elements of the <see cref="T:System.Collections.Generic.ICollection`1" /> to an <see cref="T:System.Array" />, starting at a particular <see cref="T:System.Array" /> index.</summary>
        /// <param name="array">The one-dimensional <see cref="T:System.Array" /> that is the destination of the elements copied from <see cref="T:System.Collections.Generic.ICollection`1" />. The <see cref="T:System.Array" /> must have zero-based indexing.</param>
        /// <param name="arrayIndex">The zero-based index in <paramref name="array" /> at which copying begins.</param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="array" /> is null.</exception>
        /// <exception cref="T:System.ArgumentOutOfRangeException">
        /// <paramref name="arrayIndex" /> is less than 0.</exception>
        /// <exception cref="T:System.ArgumentException">The number of elements in the source <see cref="T:System.Collections.Generic.ICollection`1" /> is greater than the available space from <paramref name="arrayIndex" /> to the end of the destination <paramref name="array" />.</exception>
        public void CopyTo(KeyValuePair<object, object>[] array, int arrayIndex)
        {
            throw new NotImplementedException();
        }

        /// <summary>Removes the first occurrence of a specific object from the <see cref="T:System.Collections.Generic.ICollection`1" />.</summary>
        /// <returns>true if <paramref name="item" /> was successfully removed from the <see cref="T:System.Collections.Generic.ICollection`1" />; otherwise, false. This method also returns false if <paramref name="item" /> is not found in the original <see cref="T:System.Collections.Generic.ICollection`1" />.</returns>
        /// <param name="item">The object to remove from the <see cref="T:System.Collections.Generic.ICollection`1" />.</param>
        /// <exception cref="T:System.NotSupportedException">The <see cref="T:System.Collections.Generic.ICollection`1" /> is read-only.</exception>
        public bool Remove(KeyValuePair<object, object> item)
        {
            throw new NotImplementedException();
        }

        /// <summary>Determines whether the <see cref="T:System.Collections.Generic.Dictionary`2" /> contains the specified key.</summary>
        /// <returns>true if the <see cref="T:System.Collections.Generic.Dictionary`2" /> contains an element with the specified key; otherwise, false.</returns>
        /// <param name="key">The key to locate in the <see cref="T:System.Collections.Generic.Dictionary`2" />.</param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="key" /> is null.</exception>
        public bool ContainsKey(object key)
        {
            return _allProperties.ContainsKey(key);
        }

        /// <summary>Determines whether the <see cref="T:System.Collections.Generic.Dictionary`2" /> contains a specific value.</summary>
        /// <returns>true if the <see cref="T:System.Collections.Generic.Dictionary`2" /> contains an element with the specified value; otherwise, false.</returns>
        /// <param name="value">The value to locate in the <see cref="T:System.Collections.Generic.Dictionary`2" />. The value can be null for reference types.</param>
        public bool ContainsValue(object value)
        {
            return _allProperties.ContainsValue(value);
        }


        /// <summary>Implements the <see cref="T:System.Runtime.Serialization.ISerializable" /> interface and returns the data needed to serialize the <see cref="T:System.Collections.Generic.Dictionary`2" /> instance.</summary>
        /// <param name="info">A <see cref="T:System.Runtime.Serialization.SerializationInfo" /> object that contains the information required to serialize the <see cref="T:System.Collections.Generic.Dictionary`2" /> instance.</param>
        /// <param name="context">A <see cref="T:System.Runtime.Serialization.StreamingContext" /> structure that contains the source and destination of the serialized stream associated with the <see cref="T:System.Collections.Generic.Dictionary`2" /> instance.</param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="info" /> is null.</exception>
        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            _allProperties.GetObjectData(info, context);
        }

        /// <summary>Implements the <see cref="T:System.Runtime.Serialization.ISerializable" /> interface and raises the deserialization event when the deserialization is complete.</summary>
        /// <param name="sender">The source of the deserialization event.</param>
        /// <exception cref="T:System.Runtime.Serialization.SerializationException">The <see cref="T:System.Runtime.Serialization.SerializationInfo" /> object associated with the current <see cref="T:System.Collections.Generic.Dictionary`2" /> instance is invalid.</exception>
        public void OnDeserialization(object sender)
        {
            _allProperties.OnDeserialization(sender);
        }

        /// <summary>Removes the value with the specified key from the <see cref="T:System.Collections.Generic.Dictionary`2" />.</summary>
        /// <returns>true if the element is successfully found and removed; otherwise, false.  This method returns false if <paramref name="key" /> is not found in the <see cref="T:System.Collections.Generic.Dictionary`2" />.</returns>
        /// <param name="key">The key of the element to remove.</param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="key" /> is null.</exception>
        public bool Remove(object key)
        {
            return _allProperties.Remove(key);
        }

        /// <summary>Gets the value associated with the specified key.</summary>
        /// <returns>true if the <see cref="T:System.Collections.Generic.Dictionary`2" /> contains an element with the specified key; otherwise, false.</returns>
        /// <param name="key">The key of the value to get.</param>
        /// <param name="value">When this method returns, contains the value associated with the specified key, if the key is found; otherwise, the default value for the type of the <paramref name="value" /> parameter. This parameter is passed uninitialized.</param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="key" /> is null.</exception>
        public bool TryGetValue(object key, out object value)
        {
            return _allProperties.TryGetValue(key, out value);
        }

        /// <summary>Gets the <see cref="T:System.Collections.Generic.IEqualityComparer`1" /> that is used to determine equality of keys for the dictionary. </summary>
        /// <returns>The <see cref="T:System.Collections.Generic.IEqualityComparer`1" /> generic interface implementation that is used to determine equality of keys for the current <see cref="T:System.Collections.Generic.Dictionary`2" /> and to provide hash values for the keys.</returns>
        public IEqualityComparer<object> Comparer => _allProperties.Comparer;

        /// <summary>Gets the number of key/value pairs contained in the <see cref="T:System.Collections.Generic.Dictionary`2" />.</summary>
        /// <returns>The number of key/value pairs contained in the <see cref="T:System.Collections.Generic.Dictionary`2" />.</returns>
        public int Count => _allProperties.Count;

        /// <summary>Gets a value indicating whether the <see cref="T:System.Collections.Generic.ICollection`1" /> is read-only.</summary>
        /// <returns>true if the <see cref="T:System.Collections.Generic.ICollection`1" /> is read-only; otherwise, false.</returns>
        public bool IsReadOnly { get; }


        /// <summary>Gets or sets the value associated with the specified key.</summary>
        /// <returns>The value associated with the specified key. If the specified key is not found, a get operation throws a <see cref="T:System.Collections.Generic.KeyNotFoundException" />, and a set operation creates a new element with the specified key.</returns>
        /// <param name="key">The key of the value to get or set.</param>
        /// <exception cref="T:System.ArgumentNullException">
        /// <paramref name="key" /> is null.</exception>
        /// <exception cref="T:System.Collections.Generic.KeyNotFoundException">The property is retrieved and <paramref name="key" /> does not exist in the collection.</exception>
        public object this[object key]
        {
            get => _allProperties[key];
            set => _allProperties[key] = value;
        }

        /// <summary>Gets an <see cref="T:System.Collections.Generic.ICollection`1" /> containing the keys of the <see cref="T:System.Collections.Generic.IDictionary`2" />.</summary>
        /// <returns>An <see cref="T:System.Collections.Generic.ICollection`1" /> containing the keys of the object that implements <see cref="T:System.Collections.Generic.IDictionary`2" />.</returns>
        public ICollection<object> Keys { get; }

        /// <summary>Gets an <see cref="T:System.Collections.Generic.ICollection`1" /> containing the values in the <see cref="T:System.Collections.Generic.IDictionary`2" />.</summary>
        /// <returns>An <see cref="T:System.Collections.Generic.ICollection`1" /> containing the values in the object that implements <see cref="T:System.Collections.Generic.IDictionary`2" />.</returns>
        public ICollection<object> Values { get; }

        #endregion

        #region Implementation of IProperties

        /// <summary>
        /// Add propery to the end of the collection 
        /// </summary>
        /// <param name="key"></param>
        /// <param name="value"></param>
        public void AddOrderedPropery(string key, object value)
        {
            _allProperties[key] = value;
            _propertyIndexes[key] = _propertyIndexes.Count;
        }

        /// <summary>
        /// Add properies to the end of the collection 
        /// </summary>
        /// <param name="properties"></param>
        public void AddOrderedProperies(IEnumerable<KeyValuePair<string, object>> properties)
        {
            foreach (var property in properties)
            {
                AddOrderedPropery(property.Key, property.Value);
            }
        }

        /// <summary>
        /// Remove ordered property by key
        /// </summary>
        /// <param name="key"></param>
        /// <returns>found?</returns>
        public bool RemoveOrderedProperty(string key)
        {
            var removed = _allProperties.Remove(key);

            if (removed)
            {
                _propertyIndexes.Remove(key);
            }
            return removed;
        }

        /// <summary>
        /// Get ordered properties in order of the template
        /// </summary>
        /// <returns></returns>
        public IList<object> GetOrderedProperties()
        {
            var arr = new object[_propertyIndexes.Count];
            foreach (var prop in _propertyIndexes)
            {
                arr[prop.Value] = _allProperties[prop.Value];
            }

            return arr;
        }

        /// <summary>
        /// Get all properties which aren't in the template
        /// </summary>
        /// <returns></returns>
        public IDictionary<object, object> GetNonOrderedProperties()
        {
            var dict = new Dictionary<object,object>();

            foreach (var prop in _allProperties)         {
                if (!_propertyIndexes.ContainsKey(prop.Value)) //todo handle string?
                {
                    dict.Add(prop.Key, prop.Value);
                }
            }
            return dict;
        }

        #endregion
    }
```

notes:

- I renamed it to orderedProperty, but maybe templateProperty is better. 
- I've generated a lot of the delegating code with Resharper (haven't checked it yet)
- I'm not sure sure if the orderedProperties names are always strings, anyway this gives issues in `GetNonOrderedProperties` (see code)
- if I'm missing something, please let me know
- feel free to edit / counter propose. Your expertise on memory usage is really appreciated! (i'm more a code complexity guy)

snakefoot commented:
Will this new IEventProperties-interface have a new LogEventInfo-property, and the old LogEventInfo-property will just point to this one?

snakefoot commented:
>The idea is that LogEventInfo.Property type will be changed from IDictionary<object, object> to IProperties. As this property is getter only, it's backwards comp.

Was think it would cause issues, but if not then fine with me.

I have now implemented the new interface and removed LogEventInfo.MessageTemplateProperties

304NotModified commented on tests/NLog.UnitTests/LogMessageFormatterTests.cs:
```diff
@@ -34,21 +34,22 @@
-            logEventInfo.Parameters = new object[] { "Login request from John for BestApplicationEver" };   // Must have parameter to activate MessageFormatter
+            foreach (var item in new KeyValuePair<string, object>[] { new KeyValuePair<string, object>("Username", "John"), new KeyValuePair<string, object>("Application", "BestApplicationEver") })
+                logEventInfo.Properties.AddOrderedProperty(item.Key, item.Value);
+            logEventInfo.Parameters = new object[] { "John", "BestApplicationEver", "Login request from John for BestApplicationEver" };
```
Why do we need this beside the `logEventInfo.Properties`? I think this unit test is doing to much, as I'm confused. Are we testing multiple things here?

snakefoot replied:
For normal structured logging you get a raw-message and arguments-params. Then based on that you derive the FormattedMessage and fill Properties-dictionary. In the Microsoft Extension Logging world you get a pre-formatted-message and can dig out some message-properties (where the last property is actually the original raw-message). For the Microsoft Extension Logging, then I want to capture the raw-message, capture the microsoft-formatted-message (store it as arguments-params), and fill the Properties-dictionary.

304NotModified commented on src/NLog/IEventProperties.cs:
```diff
@@ -0,0 +1,63 @@
+        /// Get ordered properties in order of the template
+        /// </summary>
+        /// <returns></returns>
+        IEnumerable<KeyValuePair<string, object>> OrderedProperties { get; }
```
I'm doubting of ` IList<KeyValuePair<string, object>>` instead of ` IEnumerable<KeyValuePair<string, object>>`.  What do you think is best? (note that it's a getter only)

snakefoot replied:
Well the IList-interface have an Add-method. Not sure I want to support that.

snakefoot commented:
> The idea is that LogEventInfo.Property type will be changed from IDictionary<object, object> to IProperties. As this property is getter only, it's backwards comp.

Think it is only compatible when doing a recompile, when just replacing an old NLog-assembly with the new that have changed the return-type of the Properties-property, then one get this exception:

```
Unhandled Exception: System.MissingMethodException: Method not found: 'System.Collections.Generic.IDictionary`2<System.Object,System.Object> NLog.LogEventInfo.get_Properties()'.
   at TestApplication.Program.Main(String[] args)
```

This is my advanced test application:
```
        static void Main(string[] args)
        {
            var hello = new NLog.LogEventInfo();
            hello.Properties["Hello"] = "Goodbye";
            Console.ReadKey();
        }
```

304NotModified commented:
> Think it is only compatible when doing a recompile, when just replacing an old NLog-assembly with the new that have changed the return-type of the Properties-property, then one get this exception:

Oops, binary breaking change I think. See https://github.com/dotnet/corefx/blob/c02d33b18398199f6acc17d375dab154e9a1df66/Documentation/coding-guidelines/breaking-change-rules.md#signatures

Not sure if we could fix this by *also* adding `IDictionary<object, object> LogEventInfo.Properties` and implicit interface implementation (we need also a new interface I think)

snakefoot commented:
> Oops, binary breaking change I think. 

Just like I was expecting. Guess we need to keep the LogEventInfo.Properties unchanged. Now we just have to figure out how this OrderedProperties should work.

- Should it only return the properties marked as ordered ?
- Should it just be an IList or be a custom interface that by default return all properties, but has a property to return the subset of ordered-properties?

304NotModified commented:
> Just like I was expecting. Guess we need to keep the LogEventInfo.Properties unchanged.

Or considering moving to NLog 5. (well replacing NLog 5's current state with this)

snakefoot commented:
> Or considering moving to NLog 5.

The only reason for getting this OrderedProperties or MessageTemplateProperties or NamedParameters is to have an interface to correctly capture the structured logging details from other frameworks like Microsoft Extension Logging. And maybe also use the interface to correctly forward the structured logging details to another logging framework (can see that logging frameworks often have adapters to each other).

Still confused why we can have LogEventInfo.Parameters and LogEventInfo.Properties, but we cannot have LogEventInfo.NamedParameters.

304NotModified commented:
> Still confused why we can have LogEventInfo.Parameters and LogEventInfo.Properties, but we cannot have LogEventInfo.NamedParameters (optimized to point to the same PropertiesDictionary).

Is it a NamedParameters? Or NamedButAlsoOrderedParameters? The latter is really a bit confusing. 

> Maybe time for Plan B - MessageTemplate-object. Then Microsoft Extension Logging will just have to pay the penalty of allocating the MessageTemplate-object to correctly forward the structured-logging-details.

Is the MessageTemplate as class or struct? Is the allocating penalty high?

snakefoot commented:
> Is the MessageTemplate as class or struct? Is the allocating penalty high?

Will probably be the same as allocating a List-object with an internal array of hole-structs (Name + Value + Format + possible message-string-indexof). Will probably be an interface, that we will implemented by the PropertiesDictionary.

I'm guessing you would have a better idea of this interface, since you created NLog.StructuredEvents that I have based the logic upon (So think the Template-object but without the Literals-array)

snakefoot commented:
@304NotModified Now changed the PropertiesDictionary to also implement IList (that only returns the positional properties, and added LogEventInfo.NamedParameters.

304NotModified commented:
So a namedparameter is a kind of property? I have difficulties to explain that to someone. 

Counter proposal: drop namedparameter property and add a method: GetPropertiesDefinedInTemplate (work name) that returns the IList, or maybe event better a new IOrdererDictionary (custom made interface). Maybe we need also an "add" method then.

snakefoot commented:
> So a namedparameter is a kind of property? I have difficulties to explain that to someone.

As easy to explain as MessageTemplate is also kind of a property-collection :). Actually just noticed that Serilog doesn't ensure the Properties-collection is sorted (though it usually is), but Serilog LogEvent.Properties automatically includes the properties in MessageTemplate.

304NotModified commented:
> As easy to explain as MessageTemplate is also kind of a property-sub-collection :).

true, but it not visible now from the code. 2 properties that are changing the same stuff doesn't look right.

snakefoot commented:
> true, but it not visible now from the code. 2 properties that are changing the same stuff doesn't look right.

But you want this. Since you want the Properties to include the MessageTemplate-properties.

304NotModified commented:
All I want is 2 kind properties on the logeventinfo (named and unnamed), not 3.

snakefoot commented:
> All I want is 2 kind properties on the logeventinfo (named and unnamed), not 3.

Lets make the MessageTemplate then. Then Seq will also be very happy.

304NotModified commented:
but that one was slow?

snakefoot commented:
> but that one was slow?

Can make something that is fast. Will work just like now then instead of IList then it will be IMessageTemplate.

304NotModified commented:
So

Adding `LogEventInfo.MessageTemplate` with type `IMessageTemplate`, correct?

Will `IMessageTemplate` have the full template or is it the `IList` of KeyValue pairs?

snakefoot commented:
> Will IMessageTemplate have the full template or is it the IList of KeyValue pairs?

Something like this:

```
    struct NamedParameter
    {
        public readonly string Name;
        public readonly string Format;
        public readonly object Value;
    }

    interface IMessageTemplate
    {
        string Message { get; }
        bool Positional { get; }
        IList<NamedParameter> NamedParameters { get; }
    }
```

304NotModified commented:
What about (only names changed):

```c#
    struct TemplateParameter
    {
        public readonly string Name;
        public readonly string Format;
        public readonly object Value;
    }

    interface IMessageTemplate
    {
        string Template { get; }
        bool Positional { get; }
        IList<TemplateParameter> Parameters{ get; }
    }
```

snakefoot commented:
> What about (only names changed):

Works for me. Should we keep IList or make it IEnumerable ? Can make it implicit readonly by throwing not supported on Add, Remove, etc.

304NotModified commented:
> Works for me. Should we keep IList or make it IEnumerable ?

Good one. The `IEnumerable ` is really small and throwing runtime exceptions is bad. What about creating a `NLog.IReadonlyList` (which inherits from `IEnumerable `)?

snakefoot commented:
Else we can make it like this:

```
interface IMessageTemplateParameterList : IEnumerable<MessageTemplateParameter>
{
        MessageTemplateParameter this[int index] { get; }
        int Count { get; }
}

```
And just have this property on LogEventInfo:

`IMessageTemplateParameterList  MessageTemplateParameterList { get; set; }`

304NotModified commented:
I was thinking now:

```c#
class LogEventInfo 
{
   ...
  IMessageTemplate MessageTemplate {get;}
}

interface IMessageTemplateParameterList : IEnumerable<MessageTemplateParameter>
{
        MessageTemplateParameter this[int index] { get; }
        int Count { get; }
}


    struct TemplateParameter
    {
        public readonly string Name;
        public readonly string Format;
        public readonly object Value;
    }

    interface IMessageTemplate
    {
        string Template { get; }
        bool Positional { get; }
        IMessageTemplateParameterList Parameters{ get; }
    }

```

snakefoot commented:
I have made some small hacks that might need extra scrutiny:

- LogEventInfo implements IMessageTemplate (as it already has the Message-property).
- MessageTemplate is an example of how to implement the IMessageTemplate. It inherits directly from List to minimize allocation. Right now only used for testing, but maybe others will find it useful (Like NLog Extension Logging).

304NotModified commented:
> LogEventInfo implements IMessageTemplate 

Is this needed (for performance?), its a bit strange. I would remove it if not needed. Also, is the message of the logevent the same as the template's message? Isn't the first one rendered?

snakefoot commented:
>> LogEventInfo implements IMessageTemplate
>
> Is this needed (for performance?), its a bit strange. I would remove it if not needed. 

After a good nights sleep, then I have removed this oddity, and changed it to an allocation of MessageTemplateReadOnly

>  Also, is the message of the logevent the same as the template's message? Isn't the first one rendered?

LogEventInfo.FormattedMessage is one already rendered. LogEventInfo.Message is the raw message. MessageTemplate.Message is the raw messsage.

304NotModified commented on src/NLog/IMessageTemplate.cs:
```diff
@@ -0,0 +1,144 @@
+        /// <summary>
+        /// 
+        /// </summary>
+        public IMessageTemplateParameterList Parameters { get { return this; } }
```
Is this for performance reasons? It's a bit strange design to have a property that is the same as `this` IMO

snakefoot replied:
It is to minimize allocations and because we are not using IReadOnlyList

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,744 @@
+            public PropertyValue(object value, bool messageProperty)
+            {
+                Value = value;
+                MessageProperty = messageProperty;
```
What's messageproperty? The name or the value, or something else?

snakefoot replied:
Means the dictionary-entry comes from the _messageProperty-list.

304NotModified replied:
Please rename to IsMessageProperty

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,744 @@
+                if (_eventProperties == null && value.Count <= 10)
+                {
+                    bool uniqueMessageProperties = true;
+                    for (int i = 0; i < value.Count - 1; ++i)
```
What's uniqueMessageProperties  here? What are we trying to do here?

snakefoot replied:
Ensure the MessageTemplate-parameters can be used as Dictionary-values. Ex. "Male user: {userid} loves female user: {userid}".

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,744 @@
+        void ICollection.CopyTo(Array array, int arrayIndex)
+        {
+            if (array == null)
+                throw new ArgumentNullException("array");
```
please use `nameof` :)

snakefoot replied:
Still not used to having the C# 6.0 compiler after changing to VS2017

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,744 @@
+
+        void ResetMessageProperties()
+        {
+            if (_eventProperties != null && _messageProperties != null && _messageProperties.Count > 0)
```
is this `if(IsEmpty)`?

snakefoot replied:
Will change it to IsEmpty (with some extra)

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,744 @@
+            }
+        }
+
+        private static bool InsertMessagePropertiesDictionary(IList<MessageTemplateParameter> messageProperties, Dictionary<object, PropertyValue> eventProperties)
```
for the docs phase: please document bool return values

snakefoot replied:
Please wait with doc-stuff until we have something real

304NotModified commented on src/NLog/LogEventInfo.cs:
```diff
@@ -281,20 +280,75 @@ public string FormattedMessage
+                    messageTemplateParameters = this.properties.MessageProperties;
+                    if (this.properties.MessageProperties[0].Name != null)
+                    {
+                        if (this.properties.MessageProperties[0].Name.Length != 1 || this.properties.MessageProperties[0].Name[0] < '0' || this.properties.MessageProperties[0].Name[0] > '9')
```
why is this needed in the getter? For performance reasons?

snakefoot replied:
Will remove it, since it will never happen

snakefoot replied:
Fixed

304NotModified commented on tests/NLog.UnitTests/LogMessageFormatterTests.cs:
```diff
@@ -34,21 +34,26 @@
+                new MessageTemplateParameter("Username", "John", null),
+                new MessageTemplateParameter("Application", "BestApplicationEver", null)
+            });
+            logEventInfo.Parameters = new object[] { "John", "BestApplicationEver", "Login request from John for BestApplicationEver" };
```
please split the unit test into 2 tests, or don't use the same values (john, BestApplicationEver).

snakefoot replied:
The unit test displays how I'm planning capture the structured parameters from Microsoft Extension Logging.

304NotModified replied:
This test is a bit confusing because duplicate values  (john, BestApplicationEver). Is this really needed?  What if we forgot one of the two?

304NotModified commented:
Also, do we still need the propertiesdictionary (and interface)? It was needed with the 2 properties on logeventinfo, but now we have 3 parameters.

snakefoot commented:
> Also, do we still need the propertiesdictionary (and interface)?

PropertiesDictionary ensures LogEventInfo.Properties includes the parameters from LogEventInfo.MessageTemplate.

304NotModified commented:
> Have been jumping through many hoops so far, without anything being decided yet.

Sorry, it's hard to discuss things, slow-chatting and while the code evolves all the time.  I prefer also a decision before building. 

> Reason why I have now reverted to introducing LogEventInfo.MessageTemplate was that you selected this solution yourself in NLog 4.5 Alpha. 

I think that's the best design when the perf is good enough. 

> Was hoping we could continue on and stop the interface-fight, and focus on implementation.

In my opinion that's just i'm trying to do currently. 

I think we not on the same line now. Sorry about that. What do we think we still have to decide?

snakefoot commented:
> I think we not on the same line now. Sorry about that. What do we think we still have to decide?

Well comments like "we still need the propertiesdictionary?", usually make me think that we are still stuck in the decision phase :)

> LogEventInfo.MessageTemplate - I think that's the best design when the perf is good enough.

Well I think the performance of this PR is good. Maybe considering to provide internal access to the actual LogEventInfo.PropertiesDictionary and change its enumerator to struct. Then the enumeration would be allocation free for the internal NLog-logic. But that optimization is for a different day.

I cheated in #2208 where I dropped the MessageTemplate completely, because of its overhead. Now the current interface allows me to have it without a serious performance hit.

304NotModified commented:
ps updated my reaction ;)

304NotModified commented:
> Also like simple code, but I don't like open questions about design without any counter-suggestion, then one is left in the dark trying to find the light-switch.

Fair enough. But keep in mind it's hard to give good suggestions if code is submitted which isn't clear from itself. Don't get me wrong, but it takes a lot of time to understand some parts and i'm currently have to limited time for that.

Maybe it's a better idea, that would work better for me, to create a PR with only non-optimized code (just a clear API, small changes), and in another PR a (memory optimized) change.   Design is now mixed with (highy)optimized changes.

snakefoot commented:
> Maybe it's a better idea, that would work better for me, to create a PR with only non-optimized code 

The logic that ensures LogEventInfo.Properties is a combined dictionary of both, and allows replacing of the MessageTemplate.Parameters is quite complicated. Not sure I can make it less complicated.

If you just want to see the new public API, then you just have to look at LogEventInfo.MessageTemplate. Everything else is the same.

You are welcome to present your own simple implementation of the LogEventInfo.MessageTemplate and its interaction with LogEventInfo.Properties.

304NotModified commented:
hi @snakefoot 

I thought about this and checked the code more in-depth.

In my opinion it's a bit strange that the LogEventInfo.MessageTemplate has a setter. I think this is a result of earlier dissensions (probably mine ;) )
Also I really like the idea behind the PropertyDictionary but as we can't use it well without breaking changes (no `PropertiesDictionary Properties` in LogEventInfo, I;m doubting if we really need it in a new impl.

I have a different, out-of-the-box, proposal, please let me know what you think:

Only add to the LogEventInfo:

     public IPropertiesMetaInfo PropertiesMetaInfo { get; }

Which holds the extra info which is now in the PropertiesDictionary 

The interface:

```c#
    public interface IPropertiesMetaInfo : IDictionary<string, IPropertyMetaInfo>
    {
        
    }

    public interface IPropertyMetaInfo
    {
        int Position {get;}

        string Format { get; }
    }

```

Downside is that it's a bit more work for the client to enumerate all properties in the template, but we could add a helper (dunno where).

If you think the current design is better, I really like to know!

snakefoot commented:
I don't understand what you are trying to solve with your proposal besides making Nlog more difficult to use (dictionary interface is not simple).

The reason why I have a MessageTemplate setter is because of Microsoft Extensions Logging. We can remove the setter by creating a new LogEventInfo constructor. Then we don't have the need to support only removing the messagetemplate properties. This will simplify the logic, but not get rid of PropertiesDictionary as I want the ordered guarantee



Sent from my Samsung device


-------- Original message --------
From: Julian Verdurmen <notifications@github.com>
Date: 30/09/2017 23:14 (GMT+01:00)
To: NLog/NLog <NLog@noreply.github.com>
Cc: Rolf Kristensen <sweaty1@hotmail.com>, Mention <mention@noreply.github.com>
Subject: Re: [NLog/NLog] LogEventInfo.MessageTemplate - Subset of   LogEventInfo.Properties (#2262)


hi @snakefoot<https://github.com/snakefoot>

I thought about this and checked the code more in-depth.

In my opinion it's a bit strange that the LogEventInfo.MessageTemplate has a setter. I think this is a result of earlier dissensions (probably mine ;) )
Also I really like the idea behind the PropertyDictionary but as we can't use it well without breaking changes (no PropertiesDictionary Properties in LogEventInfo, I;m doubting if we really need it in a new impl.

I have a different, out-of-the-box, proposal, please let me know what you think:

Only add to the LogEventInfo:

 public IPropertiesMetaInfo PropertiesMetaInfo { get; }


Which holds the extra info which is now in the PropertiesDictionary

The interface:

    public interface IPropertiesMetaInfo : IDictionary<string, IPropertyMetaInfo>
    {

    }

    public interface IPropertyMetaInfo
    {
        int Position {get;}

        string Format { get; }
    }


Downside is that it's a bit more work for the client to enumerate all properties in the template, but we could add a helper (dunno where).

If you think the current design is better, I really like to know!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/NLog/NLog/pull/2262#issuecomment-333336327>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AK-fnGaLMkoWhGCbMeFqeXK0QE0wu8Xvks5snq8YgaJpZM4PAffU>.

snakefoot commented:
@304NotModified I have now removed the ability to modify the MessageTemplateParameters after they have been initialized, by using a LogEventInfo-constructor. And now it is only possible to access the MessageTemplateParameters if using NET4_5 or NETSTANDARD (Because of IReadOnlyList)

304NotModified commented:
> I don't understand what you are trying to solve with your proposal besides making Nlog more difficult to use (dictionary interface is not simple).

As for the API design, it's then not clear that the templateParameters is a subset of properties.

The perfect design as we didn't need any changes to the logeventinfo, and use the logeventinfo properties. This was not the case for the position number and so I have created a new property, the metainfo (in earlier design I added it the the PropertiesDictionary, but that wasn't backwardscomp too bad).

snakefoot commented:
> As for the API design, it's then not clear that the templateParameters is a subset of properties.

Please review my new proposal. I think it satisfies your requirement of not having several ways to modify LogEventInfo.Properties.

304NotModified commented on src/NLog/LogEventInfo.cs:
```diff
@@ -281,16 +296,47 @@ public string FormattedMessage
+                            case 9: parameterName = "9"; break;
+                            default: parameterName = i.ToString(); break;
+                        }
+                        messageParameters[i] = new MessageTemplateParameter(parameterName, parameters[i], null);
```
would not it been better to use `yield return` here? (depends on the return interface)

snakefoot replied:
I like for-loops better, when not needing the state-machine.

304NotModified commented on src/NLog/LogEventInfo.cs:
```diff
@@ -281,16 +296,47 @@ public string FormattedMessage
+        /// <summary>
+        /// Gets the named parameters extracted from parsing <see cref="Message"/> as MessageTemplate
+        /// </summary>
+        public IEnumerable<MessageTemplateParameter> MessageTemplateParameters
```
Could we go again to our custom interface? Even if it's only implementing `IEnumerable<MessageTemplateParameter>`, it's more flexible for the future.

snakefoot replied:
Fixed

snakefoot commented:
@304NotModified Have squashed the commits, and made the suggested changes.

Had to re-introduce the DictionaryAdapter, as I could not ensure that it would return an enumerator of DirectoryEntry-objects.

304NotModified commented on src/NLog/Internal/MessageTemplateParameters.cs:
```diff
@@ -0,0 +1,100 @@
+                    string parameterName;
+                    switch (i)
+                    {
+                        case 0: parameterName = "0"; break;
```
this is to prevent boxing, isn't?

snakefoot replied:
No to prevent allocation of string-objects.

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,642 @@
+
+        public IDictionary EventContext { get { return _eventContextAdapter ?? (_eventContextAdapter = new DictionaryAdapter<object, object>(this)); } }
+
+        private Dictionary<object, PropertyValue> EventProperties
```
These are the regular properties, isn;t?

snakefoot replied:
Not sure I understand the question.

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,642 @@
+    /// The <see cref="MessageProperties" /> are returned as the first items
+    /// in the collection, and in positional order.
+    /// </summary>
+    internal sealed class PropertiesDictionary : IDictionary<object, object>
```
a sugesstion,

Is see a lot of special handling if `isEmpty`. I assume that is for more efficient memory usage.

Would a proxy design pattern be a good candidate here?

snakefoot replied:
Yes the IsEmpty checks is to prevent extra allocations, when operating on an empty-dictionary.

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,642 @@
+                    if (_messageProperties != null && _messageProperties.Count > 0)
+                    {
+                        _eventProperties = new Dictionary<object, PropertyValue>(_messageProperties.Count);
+                        if (!InsertMessagePropertiesIntoEmptyDictionary(_messageProperties, _eventProperties))
```
Why do we need some more code if `InsertMessagePropertiesIntoEmptyDictionary` = false?

snakefoot replied:
Because then there are duplicate names, and _messageProperties must be fixed.

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,642 @@
+            catch (ArgumentException)
+            {
+                // Duplicate keys found, lets try again
+                for (int i = 0; i < messageProperties.Count; ++i)
```
Could you explain me how remove is a try-again for add?

snakefoot replied:
Have fixed the comment

304NotModified commented on src/NLog/Internal/PropertiesDictionary.cs:
```diff
@@ -0,0 +1,642 @@
+                eventProperties[messageProperties[i].Name] = new PropertyValue(messageProperties[i].Value, true);
+                if (messagePropertiesUnique != null)
+                {
+                    messagePropertiesUnique.Add(messageProperties[i]);
```
I think we could simplify this code by yielding messageProperties (messageProperties[i] and messageProperties[y], what do you think?

snakefoot replied:
Never been a big fan of yield because of the state-machine it allocates.

304NotModified commented:
merged! If you have time to response to my comments, that would be great!
